### PR TITLE
This feature allows to index points, circles and boxes in PREFIX_TREE mapping for ES.

### PR DESCRIPTION
**Pull request by davidclement90**
I also fix some indentation issue.
https://github.com/JanusGraph/janusgraph/issues/293

Signed-off-by: David Clement <david.clement90@laposte.net>

### CODE REVIEW

File: janusgraph-test/src/main/java/org/janusgraph/diskstorage/indexing/IndexProviderTest.java [hunk 1]
```diff
@@ -424,12 +424,12 @@ private void storeTest(String... stores) throws Exception {
 
 
             //Update some data
-            add(store, "doc4", getDocument("I'ts all a big Bob", -100, 11.2, Geoshape.point(48.0, 8.0), Geoshape.line(Arrays.asList(new double[][] {{7.5, 47.5}, {8.5, 48.5}})), Arrays.asList("10", "11", "12"), Sets.newHashSet("10", "11"), Instant.ofEpochSecond(4)), true);
+            add(store, "doc4", getDocument("I'ts all a big Bob", -100, 11.2, Geoshape.point(-48.0, 8.0), Geoshape.point(-48.0, 8.0), Arrays.asList("10", "11", "12"), Sets.newHashSet("10", "11"), Instant.ofEpochSecond(4)), true);
```
Comment (sjudeng) [inline janusgraph-test/src/main/java/org/janusgraph/diskstorage/indexing/IndexProviderTest.java]: Is there a reason you changed the sign on the first point latitude?

Comment (davidclement90) [inline janusgraph-test/src/main/java/org/janusgraph/diskstorage/indexing/IndexProviderTest.java]: Yes, I want to test the reindexing process for prefix_tree in "doc1" so I change the sign on latitude to be sure that the document has really changed.
I do not want to change all tests thoroughly so I change also the sign en "doc4".

Comment (sjudeng) [inline janusgraph-test/src/main/java/org/janusgraph/diskstorage/indexing/IndexProviderTest.java]: Makes sense thanks

File: docs/searchpredicates.txt [hunk 2]
```diff
@@ -92,7 +92,7 @@ Additional data types will be supported in the future.
 
 [[geoshape]]
 === Geoshape Data Type
-The Geoshape data type supports representing a point, circle, box, line, polygon, multi-point, multi-line and multi-polygon. Index backends currently support indexing points, lines and polygons. Indexing multi-point, multi-line and multi-polygon properties has not been tested.
+The Geoshape data type supports representing a point, circle, box, line, polygon, multi-point, multi-line and multi-polygon. Index backends currently support indexing points, circles, boxes, lines and polygons. Indexing multi-point, multi-line and multi-polygon properties has not been tested.
```
Comment (sjudeng) [inline docs/searchpredicates.txt]: Can you add language to make clear that indexing circles and boxes is only supported under Elasticsearch?

Comment (davidclement90) [inline docs/searchpredicates.txt]: Why only on Elasticsearch ?
Solr is also tested. And I think that the code allow it. https://github.com/JanusGraph/janusgraph/blob/master/janusgraph-solr/src/main/java/org/janusgraph/diskstorage/solr/SolrIndex.java#L361

Comment (sjudeng) [inline docs/searchpredicates.txt]: Right, perfect! That code was added in #79 I just hadn't tested circle/box indexing. Thanks!

File: janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/ElasticSearchIndex.java [hunk 3]
```diff
@@ -468,18 +469,31 @@ private static Object convertToEsType(Object value) {
...
 
-    private static Object convertgeo(Geoshape geoshape) {
-        if (geoshape.getType() == Geoshape.Type.POINT) {
+    @SuppressWarnings("unchecked")
+    private static Object convertgeo(Geoshape geoshape, Mapping mapping) {
+        if (geoshape.getType() == Geoshape.Type.POINT && Mapping.PREFIX_TREE != mapping) {
             Geoshape.Point p = geoshape.getPoint();
             return new double[]{p.getLongitude(), p.getLatitude()};
-        } else if (geoshape.getType() != Geoshape.Type.BOX && geoshape.getType() != Geoshape.Type.CIRCLE) {
+        } else if (geoshape.getType() == Geoshape.Type.BOX) {
```
Comment (amcp) [inline janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/ElasticSearchIndex.java]: why is it OK not to catch IOException for BOX but required for CIRCLE?

File: janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/ElasticSearchIndex.java [hunk 4]
```diff
@@ -583,11 +597,13 @@ private String getAdditionScript(KeyInformation.IndexRetriever informations, Str
             switch (keyInformation.getCardinality()) {
                 case SET:
                 case LIST:
-                    script.append("ctx._source[\"" + e.field + "\"].add(" + convertToJsType(e.value, scriptLang) + ");");
+                    script.append("ctx._source[\"" + e.field + "\"].add(" + convertToJsType(e.value, scriptLang, Mapping.getMapping(keyInformation)) + ");");
```
Comment (amcp) [inline janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/ElasticSearchIndex.java]: use a String.format to reduce the number of frame switches

File: janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/ElasticSearchIndex.java [hunk 5]
```diff
@@ -583,11 +597,13 @@ private String getAdditionScript(KeyInformation.IndexRetriever informations, Str
...
                     if (hasDualStringMapping(keyInformation)) {
-                        script.append("ctx._source[\"" + getDualMappingName(e.field) + "\"].add(" + convertToJsType(e.value, scriptLang) + ");");
+                        script.append("ctx._source[\"" + getDualMappingName(e.field) + "\"].add(" + convertToJsType(e.value, scriptLang, Mapping.getMapping(keyInformation)) + ");");
```
Comment (amcp) [inline janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/ElasticSearchIndex.java]: String.format

File: janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/ElasticSearchIndex.java (hunk 4 above)
Comment (sjudeng) [inline janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/ElasticSearchIndex.java]: Switching to `String.format` may be a performance hit. I was looking up whether it was preferred over StringBuilder and results seemed to say concatenation and StringBuilder end up being similar (though StringBuilder may be better style), but String.format is 5-20 time slower than both.

1. https://kylewbanks.com/blog/java-string-concatenation-vs-stringbuilder-vs-string-format-performance
2. http://jacksondunstan.com/articles/3015
3. https://stackoverflow.com/questions/513600/should-i-use-javas-string-format-if-performance-is-important/513705#513705

Comment (amcp) [inline janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/ElasticSearchIndex.java]: @sjudeng Wow. Thank you for this important learning. StringBuilder is the way to go from now on.

File: janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/ElasticSearchIndex.java (hunk 5 above)
Comment (amcp) [inline janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/ElasticSearchIndex.java]: not an issue!

File: janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/ElasticSearchIndex.java (hunk 3 above)
Comment (amcp) [inline janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/ElasticSearchIndex.java]: bump

Comment (davidclement90) [inline janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/ElasticSearchIndex.java]: Yes it's OK.
IOException is throw by _geoshape.toMap()_

For Box, the map is constructed without using _geoshape.toMap()_.
For Circle, the map is constructed with _geoshape.toMap()_.

**Pull request merged**